### PR TITLE
A few optimizations that speedup training

### PR DESCRIPTION
**Pull request by hvoss-techfak**
Thanks for creating such a great project. 

I'm currently training on a rather large dataset (500+ GB) of audio data and have found a few optimizations that increase the pre-processing speed in the balance sampler from 10+ minutes down to ~15 seconds and changes to the sampling function to remove the for loop and use only torch functions, that made some pretty big changes for me (from 2 iterations/s to 3.5 iterations/s). 

Both changes produce exactly the same results as before.

### CODE REVIEW

**Comment by AlekseySh** [comment]
Hey, @hvoss-techfak 
thank you for the suggested changes and your interest in OML!

First, let's see if CI is green.

**Comment by hvoss-techfak** [comment]
I forgot to add the imports. One Second.

**Comment by AlekseySh** [comment]
Unfortunately, CI is not green, so, please, run tests locally (you will find the commands in the Makefile, you can also check the [Contributing guide](https://open-metric-learning.readthedocs.io/en/latest/oml/contributing.html))

**Comment by AlekseySh** [comment]
@hvoss-techfak double check linters, please

**Comment by hvoss-techfak** [comment]
The imports are now fixed, but the tests still give me an error with the "labels = torch.IntTensor(labels)", as apparently not all triplet inputs are given as integers? I'll run the tests locally and check if I can fix this

**Comment by hvoss-techfak** [comment]
I relaxed the IntTensor to a normal tensor and now all the short tests pass.

**Comment by AlekseySh** [comment]
let's try one more time

**Comment by AlekseySh** [comment]
double check linters please

**Comment by AlekseySh** [comment]
by the way, you can run command "make run_precommit", it will help to fix the linters

**Comment by AlekseySh** [comment]
@hvoss-techfak seems like the tests are fine now, so I will check the code soon 

could you run all_tests locally please? some my fail (because of environment, OS and so on), but it should be easy to understand if the are related to your changes or not

**Comment by hvoss-techfak** [comment]
All tests are now done. I had a few that failed because of my ddp setup and some import errors, but apart from that all other tests passed.

File: oml/miners/inbatch_hard_tri.py [hunk 1]
```diff
@@ -57,24 +56,35 @@ def _sample_from_distmat(distmat: Tensor, labels: List[int]) -> TTripletsIds:
...
+        dist_pos[~label_equal] = -float("inf")  # Set non-positives to -inf
+        hardest_pos_idx = torch.argmax(dist_pos, dim=1)
+
+        # Get the hardest negatives: argmin over the distance matrix where labels don't match (i.e. hardest negative)
+        dist_neg = distmat.clone()
+        dist_neg[~label_not_equal] = float("inf")  # Set non-negatives to +inf
+        hardest_neg_idx = torch.argmin(dist_neg, dim=1)
+
+        # Return anchor indices, positive indices, and negative indices
+        ids_anchor = torch.arange(batch_size, device=distmat.device)
```
Comment (DaloroAT) [inline oml/miners/inbatch_hard_tri.py:83]:
```
ids_anchor = torch.arange(batch_size, device=distmat.device)
...
return ids_anchor.cpu().tolist(), ...
```
Might be replaced to `ids_anchor = list(range(batch_size))`

File: oml/miners/inbatch_hard_tri.py [hunk 2]
```diff
@@ -57,24 +56,35 @@ def _sample_from_distmat(distmat: Tensor, labels: List[int]) -> TTripletsIds:
...
 
-            ids_anchor.append(i_anch)
-            ids_pos.append(i_pos)
-            ids_neg.append(i_neg)
+        label_equal.fill_diagonal_(False)
 
-        return ids_anchor, ids_pos, ids_neg
+        # Get the hardest positives: argmax over the distance matrix where labels match (i.e. hardest positive)
+        dist_pos = distmat.clone()
+        dist_pos[~label_equal] = -float("inf")  # Set non-positives to -inf
```
Comment (DaloroAT) [inline oml/miners/inbatch_hard_tri.py:74]: `dist_pos[~label_equal]` -> `dist_pos[label_not_equal]`

File: oml/miners/inbatch_hard_tri.py [hunk 3]
```diff
@@ -57,24 +56,35 @@ def _sample_from_distmat(distmat: Tensor, labels: List[int]) -> TTripletsIds:
...
+        hardest_pos_idx = torch.argmax(dist_pos, dim=1)
+
+        # Get the hardest negatives: argmin over the distance matrix where labels don't match (i.e. hardest negative)
+        dist_neg = distmat.clone()
+        dist_neg[~label_not_equal] = float("inf")  # Set non-negatives to +inf
```
Comment (DaloroAT) [inline oml/miners/inbatch_hard_tri.py:79]: `dist_neg[~label_not_equal]` -> `dist_neg[label_equal]`

**Review by DaloroAT — CHANGES_REQUESTED** [review]
Hey, @hvoss-techfak

Thank you for your contribution!

I left a couple of comments removing redundant code.

File: oml/miners/inbatch_hard_tri.py (hunk 3 above)
Comment (DaloroAT) [reply oml/miners/inbatch_hard_tri.py:79]: For this refactoring, this line `label_equal.fill_diagonal_(False)` should be removed 🤔

**Comment by AlekseySh** [comment]
@hvoss-techfak check out the comments please

**Comment by hvoss-techfak** [comment]
all changes are done

**Review by DaloroAT — APPROVED** [review]

**Comment by AlekseySh** [comment]
@DaloroAT thank you for review it!

**Comment by AlekseySh** [comment]
@hvoss-techfak welcome to the contributors, feel free to work on any issue you like :)

**Pull request merged**